File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/MessageLifecycle.kt
package com.flare.im.ui

enum class FlareMessageSendState { Draft, Sending, Sent, Failed }
enum class FlareMessageLifecycleDeliveryState { ServerAccepted, Delivered, PartiallyDelivered }
enum class FlareMessageReadState { Unread, PartiallyRead, Read }
enum class FlareMessageMutationState { Normal, Edited, Recalled, Deleted }
enum class FlareMessageEphemeralState { None, ReadOnce, BurnAfterRead, Expired }

/** UI-facing lifecycle projection mapped from authoritative host state. */
data class FlareMessageLifecycle(
    val transfer: FlareTransferState = FlareTransferState.Idle,
    val send: FlareMessageSendState = FlareMessageSendState.Sent,
    val delivery: FlareMessageLifecycleDeliveryState = FlareMessageLifecycleDeliveryState.ServerAccepted,
    val read: FlareMessageReadState = FlareMessageReadState.Unread,
    val mutation: FlareMessageMutationState = FlareMessageMutationState.Normal,
    val ephemeral: FlareMessageEphemeralState = FlareMessageEphemeralState.None,
) {
    val visualStatus: FlareMessageDeliveryStatus
        get() = when {
            send == FlareMessageSendState.Failed || transfer == FlareTransferState.Failed ->
                FlareMessageDeliveryStatus.Failed
            send == FlareMessageSendState.Draft || transfer == FlareTransferState.Queued ->
                FlareMessageDeliveryStatus.Pending
            send == FlareMessageSendState.Sending || transfer == FlareTransferState.Transferring ||
                transfer == FlareTransferState.Paused -> FlareMessageDeliveryStatus.Sending
            read == FlareMessageReadState.Read -> FlareMessageDeliveryStatus.Read
            delivery == FlareMessageLifecycleDeliveryState.Delivered ||
                delivery == FlareMessageLifecycleDeliveryState.PartiallyDelivered -> FlareMessageDeliveryStatus.Delivered
            else -> FlareMessageDeliveryStatus.Sent
        }
}

File: packages/ios-im-ui/Sources/FlareIMUI/Models/MessageLifecycle.swift
import Foundation

public enum FlareMessageSendState: String, Sendable { case draft, sending, sent, failed }
public enum FlareMessageLifecycleDeliveryState: String, Sendable { case serverAccepted, delivered, partiallyDelivered }
public enum FlareMessageReadState: String, Sendable { case unread, partiallyRead, read }
public enum FlareMessageMutationState: String, Sendable { case normal, edited, recalled, deleted }
public enum FlareMessageEphemeralState: String, Sendable { case none, readOnce, burnAfterRead, expired }

/// UI-facing lifecycle projection mapped from authoritative host state.
public struct FlareMessageLifecycle: Sendable {
    public let transfer: FlareTransferState
    public let send: FlareMessageSendState
    public let delivery: FlareMessageLifecycleDeliveryState
    public let read: FlareMessageReadState
    public let mutation: FlareMessageMutationState
    public let ephemeral: FlareMessageEphemeralState

    public init(
        transfer: FlareTransferState = .idle,
        send: FlareMessageSendState = .sent,
        delivery: FlareMessageLifecycleDeliveryState = .serverAccepted,
        read: FlareMessageReadState = .unread,
        mutation: FlareMessageMutationState = .normal,
        ephemeral: FlareMessageEphemeralState = .none
    ) {
        self.transfer = transfer
        self.send = send
        self.delivery = delivery
        self.read = read
        self.mutation = mutation
        self.ephemeral = ephemeral
    }

    public var visualStatus: FlareMessageDeliveryStatus {
        if send == .failed || transfer == .failed { return .failed }
        if send == .draft || transfer == .queued { return .pending }
        if send == .sending || transfer == .transferring || transfer == .paused { return .sending }
        if read == .read { return .read }
        if delivery == .delivered || delivery == .partiallyDelivered { return .delivered }
        return .sent
    }
}
